components/cli/pkg/registry/credentials/credentials.go

```diff
@@ -0,0 +1,187 @@
+	}
+	redirectUrl := url.QueryEscape("http://" + auth.CallBackHost + ":" + strconv.Itoa(codeReceiverPort) + "/auth")
+	var gitHubAuthUrl = "https://" + auth.IsHost + ":" + strconv.Itoa(auth.IsPort) +
+		"/oauth2/authorize?scope=openid&response_type=" + "code&redirect_uri=" + redirectUrl + "&client_id=" +
+		auth.SpClientId + "&fidp=google"
+
+	fmt.Println("\n", gitHubAuthUrl)
+	go func() {
+		mux := http.NewServeMux()
+		server := http.Server{Addr: httpPortString, Handler: mux}
+		//var timer *time.Timer
+		mux.HandleFunc(auth.CallBackContextPath, func(w http.ResponseWriter, r *http.Request) {
+			err := r.ParseForm()
+			if err != nil {
+				util.ExitWithErrorMessage("Error parsing the code", err)
+			}
+			code = r.Form.Get(auth.CallBackParameter)
+			ch <- code
+			if len(code) != 0 {
+				http.Redirect(w, r, "https://"+auth.IsHost+":"+strconv.Itoa(auth.IsPort)+
+					"/authenticationendpoint/auth_success.html", http.StatusSeeOther)
+			} else {
+				util.ExitWithErrorMessage("Did not receive any code", err)
+			}
+			flusher, ok := w.(http.Flusher)
+			if !ok {
+				util.ExitWithErrorMessage("Error in casting the flusher", err)
+			}
+			flusher.Flush()
+			err = server.Shutdown(context.Background())
+			if err != nil {
+				util.ExitWithErrorMessage("Error while shutting down the server\n", err)
+			}
+		})
+		if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			util.ExitWithErrorMessage("Error while establishing the service", err)
+		}
+	}()
+
+	err := util.OpenBrowser(gitHubAuthUrl)
+	if err != nil {
+		fmt.Printf("Could not resolve the given url %s. Started to operate in the headless "+
+			"mode\n", gitHubAuthUrl)
+		return FromTerminal(username)
+	}
+	// Setting up a timeout
+	go func() {
+		time.Sleep(5 * time.Minute)
+		timeout <- true
+	}()
+	// Wait for a code, or timeout
+	select {
+	case <-ch:
+	case <-timeout:
+		util.ExitWithErrorMessage("Failed to authenticate", errors.
+			New("time out. Did not receive any code"))
+	}
+	token := getTokenFromCode(code, codeReceiverPort, auth)
+	username, accessToken := getUsernameAndTokenFromJWT(token)
+	return username, accessToken, nil
+}
+
+// FromTerminal is to allow this login flow to work in headless mode
+func FromTerminal(username string) (string, string, error) {
+	var password string
+	if username == "" {
+		fmt.Print("Enter username: ")
+		_, err := fmt.Scanln(&username)
+		if err != nil {
+			util.ExitWithErrorMessage("Error reading the input username", err)
+		}
+	}
+	fmt.Print("Enter password/token: ")
+	bytePassword, err := terminal.ReadPassword(0)
+	if err != nil {
+		util.ExitWithErrorMessage("Error reading the input token", err)
+	}
+	password = strings.TrimSpace(string(bytePassword))
+	username = strings.TrimSpace(username)
+	fmt.Println()
+	return username, password, nil
+}
+
+// getUsernameAndTokenFromJWT returns the extracted subject from the JWT
+func getUsernameAndTokenFromJWT(token string) (string, string) {
```

Comment:
Please rename this method to something more meaningful. It seems like the the token parameter passed to this method is not actually a JWT (or a token), but a JSON response from the IdP